KAFKA-2478: Fix manual committing example in javadoc

shtratos opened the pull request:
Committing before inserting all records into the database
might lead to some records being lost.

I've changed the example to commit only after all records 
returned by `poll` are inserted into the database.

shtratos committed:
KAFKA-2478: Fix manual committing example in javadoc

Committing before inserting all records into the database
might lead to some records being lost.

shtratos committed:
TRIVIAL: use diamond operator <> where possible

shtratos commented:
Alternative implementation makes sure that `buffer` size
does not grow over `commitInterval`, but it's more verbose:

```
        Properties props = new Properties();
        props.put("bootstrap.servers", "localhost:9092");
        props.put("group.id", "test");
        props.put("enable.auto.commit", "false");
        props.put("auto.commit.interval.ms", "1000");
        props.put("session.timeout.ms", "30000");
        props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        props.put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props);
        consumer.subscribe(Arrays.asList("foo", "bar"));
        int commitInterval = 200;
        List<ConsumerRecord<String, String>> buffer = new ArrayList<>();
        Map<TopicPartition, Long> consumedOffsets = new HashMap<>();
        while (true) {
            ConsumerRecords<String, String> records = consumer.poll(100);
            for (ConsumerRecord<String, String> record : records) {
                buffer.add(record);
                consumedOffsets.put(new TopicPartition(record.topic(), record.partition()), record.offset());
                if (buffer.size() >= commitInterval) {
                    insertIntoDb(buffer);
                    consumer.commitSync(consumedOffsets);
                    buffer.clear();
                    consumedOffsets.clear();
                }
            }
        }
```

Here we commit every 200 records as in original example, but we have to maintain 
the consumed offsets manually, so we don't commit records that haven't been inserted
into the database yet.

hachikuji commented:
@shtratos I think the example still needs fixed in trunk, but the code has diverged. Can you rebase?

shtratos commented:
@hachikuji rebased, please review

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -200,19 +200,19 @@
- *                 buffer.clear();
- *             }
+ *         }
+ *         if (buffer.size() &gt;= commitInterval) {
```
Instead of "commitInterval," should this be something like "minBatchSize"?

shtratos replied:
I'd go for "minCommitInterval". "batch" is too generic IMO.

hachikuji replied:
I was just thinking that it seems odd to compare the buffer size against an interval. Maybe "minCommitSize"?

hachikuji replied:
Not a big deal either way, but I actually prefer "minBatchSize" since the javadoc specifically describes this as a batching example. And people usually would think about batching in terms of the size of the batch and not the size of the commit.

shtratos replied:
Reread the javadoc, "minBatchSize" does make more sense.
Sorry, it's been a while since I looked at this code.

I changed it

shtratos replied:
`commitInterval` made sense originally, because previously the interval between every 2 consecutive committed offsets was exactly 200 records.

shtratos committed:
KAFKA-2478: Rename `commitInterval` -> `minBatchSize`

It's no longer guaranteed that commit will happen exactly
 every "commitInterval" records. Now we can only say that
 at least `minBatchSize` will be inserted in the database.

hachikuji commented:
LGTM

shtratos commented:
@hachikuji thank you. Will you merge this? Do you need any other help with this PR?

guozhangwang commented:
LGTM. Merged to both trunk and 0.9.0.

shtratos commented:
Thanks!
